File: apps/api/app/models/exam.py
from datetime import date, datetime, time
from typing import TYPE_CHECKING

from sqlalchemy import Date, DateTime, ForeignKey, String, Time, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base

if TYPE_CHECKING:
    from app.models.course import Course


class Exam(Base):
    __tablename__ = "exams"

    id: Mapped[int] = mapped_column(primary_key=True)
    course_id: Mapped[int] = mapped_column(ForeignKey("courses.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(160))
    exam_date: Mapped[date] = mapped_column(Date, index=True)
    start_time: Mapped[time | None] = mapped_column(Time, nullable=True)
    location: Mapped[str | None] = mapped_column(String(80), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    course: Mapped["Course"] = relationship(back_populates="exams")

File: apps/web/src/types/exam.ts
export type Exam = { id: number; course_id: number; title: string; exam_date: string; start_time: string | null; location: string | null; created_at: string; course: { id: number; code: string; name: string; color: string } }
export type ExamInput = { course_id: number; title: string; exam_date: string; start_time: string | null; location: string | null }
